File: apps/mobile/ios/Sources/State/SessionManager.swift
//
//  SessionManager.swift
//  TchatApp
//
//  Created by Claude on 22/09/2024.
//

import Foundation
import Combine

/// Manages cross-platform session state and workspace switching
public class SessionManager: ObservableObject {

    // MARK: - Properties
    private let stateSyncManager: StateSyncManager
    private let persistenceManager: PersistenceManager
    private var cancellables = Set<AnyCancellable>()

    @Published public var currentSession: UserSession?
    @Published public var activeWorkspace: Workspace?
    @Published public var isSessionSyncing: Bool = false
    @Published public var lastSessionSync: Date?

    // Session timeout configuration
    private let sessionTimeout: TimeInterval = 30 * 60 // 30 minutes
    private var sessionTimer: Timer?

    // MARK: - Initialization
    public init(stateSyncManager: StateSyncManager, persistenceManager: PersistenceManager) {
        self.stateSyncManager = stateSyncManager
        self.persistenceManager = persistenceManager

        loadPersistedSession()
        setupSessionMonitoring()
    }

    deinit {
        sessionTimer?.invalidate()
    }

    // MARK: - Session Management

    /// Start a new session with user authentication
    public func startSession(for user: User, workspace: Workspace) async throws {
        let session = UserSession(
            id: UUID().uuidString,
            userId: user.id,
            workspaceId: workspace.id,
            platform: "ios",
            deviceId: getDeviceId(),
            startTime: Date(),
            lastActivity: Date(),
            isActive: true
        )

        currentSession = session
        activeWorkspace = workspace

        // Sync session to server
        try await syncSessionToServer(session)

        // Start session monitoring
        startSessionMonitoring()

        // Persist session locally
        persistenceManager.saveSession(session)
        persistenceManager.saveWorkspace(workspace)

        await MainActor.run {
            NotificationCenter.default.post(
                name: .sessionDidStart,
                object: session
            )
        }
    }

    /// Switch to a different workspace
    public func switchWorkspace(_ workspace: Workspace) async throws {
        guard let session = currentSession else {
            throw SessionError.noActiveSession
        }

        // Update session with new workspace
        let updatedSession = UserSession(
            id: session.id,
            userId: session.userId,
            workspaceId: workspace.id,
            platform: session.platform,
            deviceId: session.deviceId,
            startTime: session.startTime,
            lastActivity: Date(),
            isActive: true
        )

        currentSession = updatedSession
        activeWorkspace = workspace

        // Sync workspace change to server
        try await syncWorkspaceSwitch(updatedSession, workspace)

        // Persist changes
        persistenceManager.saveSession(updatedSession)
        persistenceManager.saveWorkspace(workspace)

        await MainActor.run {
            NotificationCenter.default.post(
                name: .workspaceDidChange,
                object: workspace,
                userInfo: ["previousWorkspace": activeWorkspace as Any]
            )
        }
    }

    /// Update session activity
    public func updateActivity() async {
        guard let session = currentSession else { return }

        let updatedSession = UserSession(
            id: session.id,
            userId: session.userId,
            workspaceId: session.workspaceId,
            platform: session.platform,
            deviceId: session.deviceId,
            startTime: session.startTime,
            lastActivity: Date(),
            isActive: true
        )

        currentSession = updatedSession
        persistenceManager.saveSession(updatedSession)

        // Sync activity update to server (throttled)
        Task {
            try await throttledSessionSync(updatedSession)
        }
    }

    /// End current session
    public func endSession() async throws {
        guard let session = currentSession else {
            throw SessionError.noActiveSession
        }

        let endedSession = UserSession(
            id: session.id,
            userId: session.userId,
            workspaceId: session.workspaceId,
            platform: session.platform,
            deviceId: session.deviceId,
            startTime: session.startTime,
            lastActivity: Date(),
            isActive: false
        )

        // Sync session end to server
        try await syncSessionToServer(endedSession)

        // Clear local session
        currentSession = nil
        activeWorkspace = nil
        stopSessionMonitoring()

        // Clear persisted session
        persistenceManager.clearSession()

        await MainActor.run {
            NotificationCenter.default.post(
                name: .sessionDidEnd,
                object: endedSession
            )
        }
    }

    /// Get active sessions across all platforms
    public func getActiveSessions() async throws -> [CrossPlatformSession] {
        guard let url = URL(string: "https://api.tchat.app/api/sessions/active") else {
            throw SessionError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = getAuthToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw SessionError.serverError
        }

        return try JSONDecoder().decode([CrossPlatformSession].self, from: data)
    }

    // MARK: - Private Methods

    private func loadPersistedSession() {
        if let session = persistenceManager.loadSession(),
           let workspace = persistenceManager.loadWorkspace() {

            // Check if session is still valid (not expired)
            let timeSinceLastActivity = Date().timeIntervalSince(session.lastActivity)
            if timeSinceLastActivity < sessionTimeout {
                currentSession = session
                activeWorkspace = workspace
                startSessionMonitoring()
            } else {
                // Session expired, clear it
                persistenceManager.clearSession()
            }
        }
    }

    private func setupSessionMonitoring() {
        // Monitor app lifecycle events
        NotificationCenter.default.publisher(for: UIApplication.willEnterForegroundNotification)
            .sink { [weak self] _ in
                Task {
                    await self?.updateActivity()
                }
            }
            .store(in: &cancellables)

        NotificationCenter.default.publisher(for: UIApplication.didEnterBackgroundNotification)
            .sink { [weak self] _ in
                Task {
                    await self?.updateActivity()
                }
            }
            .store(in: &cancellables)
    }

    private func startSessionMonitoring() {
        sessionTimer?.invalidate()
        sessionTimer = Timer.scheduledTimer(withTimeInterval: 60.0, repeats: true) { [weak self] _ in
            Task {
                await self?.checkSessionExpiry()
            }
        }
    }

    private func stopSessionMonitoring() {
        sessionTimer?.invalidate()
        sessionTimer = nil
    }

    private func checkSessionExpiry() async {
        guard let session = currentSession else { return }

        let timeSinceLastActivity = Date().timeIntervalSince(session.lastActivity)
        if timeSinceLastActivity >= sessionTimeout {
            try? await endSession()
        }
    }

    private func syncSessionToServer(_ session: UserSession) async throws {
        guard let url = URL(string: "https://api.tchat.app/api/sessions/sync") else {
            throw SessionError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = getAuthToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        request.httpBody = try encoder.encode(session)

        let (_, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw SessionError.syncFailed
        }

        lastSessionSync = Date()
    }

    private func syncWorkspaceSwitch(_ session: UserSession, _ workspace: Workspace) async throws {
        let payload = WorkspaceSwitchPayload(
            sessionId: session.id,
            userId: session.userId,
            newWorkspaceId: workspace.id,
            timestamp: Date(),
            platform: "ios"
        )

        guard let url = URL(string: "https://api.tchat.app/api/sessions/workspace-switch") else {
            throw SessionError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = getAuthToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        request.httpBody = try encoder.encode(payload)

        let (_, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw SessionError.syncFailed
        }
    }

    private var lastSyncTime: Date = Date.distantPast
    private func throttledSessionSync(_ session: UserSession) async throws {
        let now = Date()
        let timeSinceLastSync = now.timeIntervalSince(lastSyncTime)

        // Only sync if more than 30 seconds have passed
        if timeSinceLastSync >= 30 {
            try await syncSessionToServer(session)
            lastSyncTime = now
        }
    }

    private func getDeviceId() -> String {
        return UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
    }

    private func getAuthToken() -> String? {
        // Retrieve authentication token from keychain or secure storage
        return nil // Placeholder
    }
}

// MARK: - Session Models

public struct UserSession: Codable, Identifiable {
    public let id: String
    public let userId: String
    public let workspaceId: String
    public let platform: String
    public let deviceId: String
    public let startTime: Date
    public let lastActivity: Date
    public let isActive: Bool
}

public struct Workspace: Codable, Identifiable {
    public let id: String
    public let name: String
    public let description: String?
    public let iconUrl: String?
    public let memberCount: Int
    public let isPersonal: Bool
}

public struct CrossPlatformSession: Codable {
    public let sessionId: String
    public let platform: String
    public let deviceType: String
    public let lastActivity: Date
    public let isCurrentSession: Bool
}

public struct WorkspaceSwitchPayload: Codable {
    public let sessionId: String
    public let userId: String
    public let newWorkspaceId: String
    public let timestamp: Date
    public let platform: String
}

public enum SessionError: Error, LocalizedError {
    case noActiveSession
    case invalidURL
    case serverError
    case syncFailed
    case sessionExpired

    public var errorDescription: String? {
        switch self {
        case .noActiveSession:
            return "No active session found"
        case .invalidURL:
            return "Invalid session management URL"
        case .serverError:
            return "Session server error"
        case .syncFailed:
            return "Failed to sync session"
        case .sessionExpired:
            return "Session has expired"
        }
    }
}

// MARK: - Session Notification Extensions

extension Notification.Name {
    static let sessionDidStart = Notification.Name("sessionDidStart")
    static let sessionDidEnd = Notification.Name("sessionDidEnd")
    static let workspaceDidChange = Notification.Name("workspaceDidChange")
}

File: apps/mobile/android/app/src/main/java/com/tchat/state/SessionManager.kt
package com.tchat.state

import android.content.Context
import android.provider.Settings
import com.tchat.models.User
import com.tchat.models.Workspace
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.Contextual
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Manages cross-platform session state and workspace switching
 */
class SessionManager(
    private val context: Context,
    private val stateSyncManager: StateSyncManager,
    private val persistenceManager: PersistenceManager
) {

    // MARK: - Properties
    private val baseUrl = "https://api.tchat.app"
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    private val _currentSession = MutableStateFlow<UserSession?>(null)
    val currentSession: StateFlow<UserSession?> = _currentSession.asStateFlow()

    private val _activeWorkspace = MutableStateFlow<Workspace?>(null)
    val activeWorkspace: StateFlow<Workspace?> = _activeWorkspace.asStateFlow()

    private val _isSessionSyncing = MutableStateFlow(false)
    val isSessionSyncing: StateFlow<Boolean> = _isSessionSyncing.asStateFlow()

    private val _lastSessionSync = MutableStateFlow<Date?>(null)
    val lastSessionSync: StateFlow<Date?> = _lastSessionSync.asStateFlow()

    // Session timeout configuration
    private val sessionTimeout = 30 * 60 * 1000L // 30 minutes in milliseconds
    private var sessionMonitoringJob: Job? = null
    private var lastSyncTime = Date(0)

    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    init {
        loadPersistedSession()
    }

    // MARK: - Session Management

    /**
     * Start a new session with user authentication
     */
    suspend fun startSession(user: User, workspace: Workspace) {
        val session = UserSession(
            id = UUID.randomUUID().toString(),
            userId = user.id,
            workspaceId = workspace.id,
            platform = "android",
            deviceId = getDeviceId(),
            startTime = Date(),
            lastActivity = Date(),
            isActive = true
        )

        _currentSession.value = session
        _activeWorkspace.value = workspace

        // Sync session to server
        syncSessionToServer(session)

        // Start session monitoring
        startSessionMonitoring()

        // Persist session locally
        persistenceManager.saveSession(session)
        persistenceManager.saveWorkspace(workspace)

        // Notify session start
        notifySessionStart(session)
    }

    /**
     * Switch to a different workspace
     */
    suspend fun switchWorkspace(workspace: Workspace) {
        val session = _currentSession.value
            ?: throw SessionError.NoActiveSession

        // Update session with new workspace
        val updatedSession = session.copy(
            workspaceId = workspace.id,
            lastActivity = Date()
        )

        _currentSession.value = updatedSession
        _activeWorkspace.value = workspace

        // Sync workspace change to server
        syncWorkspaceSwitch(updatedSession, workspace)

        // Persist changes
        persistenceManager.saveSession(updatedSession)
        persistenceManager.saveWorkspace(workspace)

        // Notify workspace change
        notifyWorkspaceChange(workspace)
    }

    /**
     * Update session activity
     */
    suspend fun updateActivity() {
        val session = _currentSession.value ?: return

        val updatedSession = session.copy(
            lastActivity = Date()
        )

        _currentSession.value = updatedSession
        persistenceManager.saveSession(updatedSession)

        // Sync activity update to server (throttled)
        throttledSessionSync(updatedSession)
    }

    /**
     * End current session
     */
    suspend fun endSession() {
        val session = _currentSession.value
            ?: throw SessionError.NoActiveSession

        val endedSession = session.copy(
            lastActivity = Date(),
            isActive = false
        )

        // Sync session end to server
        syncSessionToServer(endedSession)

        // Clear local session
        _currentSession.value = null
        _activeWorkspace.value = null
        stopSessionMonitoring()

        // Clear persisted session
        persistenceManager.clearSession()

        // Notify session end
        notifySessionEnd(endedSession)
    }

    /**
     * Get active sessions across all platforms
     */
    suspend fun getActiveSessions(): List<CrossPlatformSession> = withContext(Dispatchers.IO) {
        val url = "$baseUrl/api/sessions/active"

        val requestBuilder = Request.Builder()
            .url(url)
            .get()
            .addHeader("Content-Type", "application/json")

        // Add authentication header if needed
        getAuthToken()?.let { token ->
            requestBuilder.addHeader("Authorization", "Bearer $token")
        }

        val request = requestBuilder.build()
        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw SessionError.ServerError
        }

        val responseBody = response.body?.string() ?: throw SessionError.ServerError
        json.decodeFromString<List<CrossPlatformSession>>(responseBody)
    }

    /**
     * Cleanup resources
     */
    fun cleanup() {
        stopSessionMonitoring()
        scope.cancel()
    }

    // MARK: - Private Methods

    private fun loadPersistedSession() {
        val session = persistenceManager.loadSession()
        val workspace = persistenceManager.loadWorkspace()

        if (session != null && workspace != null) {
            // Check if session is still valid (not expired)
            val timeSinceLastActivity = Date().time - session.lastActivity.time
            if (timeSinceLastActivity < sessionTimeout) {
                _currentSession.value = session
                _activeWorkspace.value = workspace
                startSessionMonitoring()
            } else {
                // Session expired, clear it
                persistenceManager.clearSession()
            }
        }
    }

    private fun startSessionMonitoring() {
        sessionMonitoringJob?.cancel()
        sessionMonitoringJob = scope.launch {
            while (isActive) {
                checkSessionExpiry()
                delay(60_000) // Check every minute
            }
        }
    }

    private fun stopSessionMonitoring() {
        sessionMonitoringJob?.cancel()
        sessionMonitoringJob = null
    }

    private suspend fun checkSessionExpiry() {
        val session = _currentSession.value ?: return

        val timeSinceLastActivity = Date().time - session.lastActivity.time
        if (timeSinceLastActivity >= sessionTimeout) {
            try {
                endSession()
            } catch (e: Exception) {
                // Handle session end error
            }
        }
    }

    private suspend fun syncSessionToServer(session: UserSession) = withContext(Dispatchers.IO) {
        _isSessionSyncing.value = true

        try {
            val url = "$baseUrl/api/sessions/sync"
            val jsonString = json.encodeToString(session)
            val mediaType = "application/json; charset=utf-8".toMediaType()
            val requestBody = jsonString.toRequestBody(mediaType)

            val requestBuilder = Request.Builder()
                .url(url)
                .post(requestBody)
                .addHeader("Content-Type", "application/json")

            // Add authentication header if needed
            getAuthToken()?.let { token ->
                requestBuilder.addHeader("Authorization", "Bearer $token")
            }

            val request = requestBuilder.build()
            val response = client.newCall(request).execute()

            if (!response.isSuccessful) {
                throw SessionError.SyncFailed
            }

            _lastSessionSync.value = Date()

        } finally {
            _isSessionSyncing.value = false
        }
    }

    private suspend fun syncWorkspaceSwitch(session: UserSession, workspace: Workspace) = withContext(Dispatchers.IO) {
        val payload = WorkspaceSwitchPayload(
            sessionId = session.id,
            userId = session.userId,
            newWorkspaceId = workspace.id,
            timestamp = Date(),
            platform = "android"
        )

        val url = "$baseUrl/api/sessions/workspace-switch"
        val jsonString = json.encodeToString(payload)
        val mediaType = "application/json; charset=utf-8".toMediaType()
        val requestBody = jsonString.toRequestBody(mediaType)

        val requestBuilder = Request.Builder()
            .url(url)
            .post(requestBody)
            .addHeader("Content-Type", "application/json")

        // Add authentication header if needed
        getAuthToken()?.let { token ->
            requestBuilder.addHeader("Authorization", "Bearer $token")
        }

        val request = requestBuilder.build()
        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw SessionError.SyncFailed
        }
    }

    private suspend fun throttledSessionSync(session: UserSession) {
        val now = Date()
        val timeSinceLastSync = now.time - lastSyncTime.time

        // Only sync if more than 30 seconds have passed
        if (timeSinceLastSync >= 30_000) {
            syncSessionToServer(session)
            lastSyncTime = now
        }
    }

    private fun getDeviceId(): String {
        return Settings.Secure.getString(
            context.contentResolver,
            Settings.Secure.ANDROID_ID
        ) ?: UUID.randomUUID().toString()
    }

    private fun getAuthToken(): String? {
        // Retrieve authentication token from secure storage
        return null // Placeholder
    }

    private fun notifySessionStart(session: UserSession) {
        // In a real app, this would use an event bus or broadcast
        // For now, this is a placeholder for session start notifications
    }

    private fun notifySessionEnd(session: UserSession) {
        // In a real app, this would use an event bus or broadcast
        // For now, this is a placeholder for session end notifications
    }

    private fun notifyWorkspaceChange(workspace: Workspace) {
        // In a real app, this would use an event bus or broadcast
        // For now, this is a placeholder for workspace change notifications
    }
}

// MARK: - Session Models

@Serializable
data class UserSession(
    val id: String,
    val userId: String,
    val workspaceId: String,
    val platform: String,
    val deviceId: String,
    val startTime: @Contextual Date,
    val lastActivity: @Contextual Date,
    val isActive: Boolean
)


@Serializable
data class CrossPlatformSession(
    val sessionId: String,
    val platform: String,
    val deviceType: String,
    val lastActivity: @Contextual Date,
    val isCurrentSession: Boolean
)

@Serializable
data class WorkspaceSwitchPayload(
    val sessionId: String,
    val userId: String,
    val newWorkspaceId: String,
    val timestamp: @Contextual Date,
    val platform: String
)

/**
 * Session errors
 */
sealed class SessionError(message: String) : Exception(message) {
    object NoActiveSession : SessionError("No active session found")
    object InvalidUrl : SessionError("Invalid session management URL")
    object ServerError : SessionError("Session server error")
    object SyncFailed : SessionError("Failed to sync session")
    object SessionExpired : SessionError("Session has expired")
}